### ISSUE TITLE
Repository mirror

### ISSUE THREAD

**Issue by dbruggner** [gitbucket/gitbucket#1556]
## Issue
**Impacted version**: 4.11

**Deployment mode**: not relevant

**Problem description**:
As a user I want a repository mirror as described in https://help.github.com/articles/duplicating-a-repository/#mirroring-a-repository-in-another-location .
When I create a new repository, I want the ability to setup a mirror. This could be a read-only repository or with forward pushing.
It should be updated automatically by a user defined interval.

In Gogs/ Gitea you can choose mirror function when migrating a repository:
![gitea_repository_mirroring](https://cloud.githubusercontent.com/assets/12123241/25444321/f86e7c90-2aaa-11e7-9315-4ac0610922a7.jpg)

**Comment by cenggel** [comment]
need this feature,too!

**Comment by kounoike** [comment]
@takezoe Is this function already implemented in 4.20.0?

**Comment by McFoggy** [comment]
@dbruggner how conflicts are handled in scenarios where teams push to both repositories?

**Comment by dbruggner** [comment]
@McFoggy I think in a first approach it's sufficient when the mirrored repositories are read-only. So that's only possible to pull from it but not pushing to it.
But it would be interesting to look at the way Gogs/ Gitea handle this scenario.

**Comment by McFoggy** [comment]
@dbruggner I think they do not care and consider just the usage of a `git push --mirror`. But even with that once we setup an automatic mechanism that pushes to a remote we also have to take care of what is going if/when it fails.

**Comment by lavvy** [comment]
Any progress on this feature? This feature is about the next big feature in gitbucket.

**Comment by aadrian** [comment]
https://github.com/alexandremenif/gitbucket-mirror-plugin

**Comment by lavvy** [comment]
Oh wow. Cool

**Comment by lavvy** [comment]
@aadrian , this is so different from the mirror feature i know. The plugin is like pushing code to a remote source and must have credentials of the remote.
i am thinking it should be how gogs , gitlab others applied the mirror feature. 
This feature still deserves to be in the upstream

**Comment by yaroot** [comment]
@lavvy so _pull mode_?

**Comment by aadrian** [comment]
@lavvy I don't get the difference. A feature like like the https://github.com/alexandremenif/gitbucket-mirror-plugin/issues/1 would do this for all repos in an GitBucket installation.

**Comment by lavvy** [comment]
Yeah @yaroot . it should be pulling instead but this plugin is pushing. 
I thought the essence of a mirror is to keep track of what is happening in a particular remote repo. But when I added a remote repo url to the mirror, this plugin ended up changing the particular repo it is supposed to watch and pull changes. 
At first, I didn't know what was happening but I later tried to observe It and found out that It is just doing the opposite of what it is supposed to do. 
I don't even know the real use of it, because it can only work on two repos that be belong to the same user, so what is the essence of keeping a sync of two repos you own.

**Comment by yaroot** [comment]
@lavvy Okay, we already have the mirror-plugin, maybe the feature belongs there, it could support both push & pull mode like how [Gitlab EE](https://docs.gitlab.com/ee/workflow/repository_mirroring.html) does.

**Comment by lavvy** [comment]
Any hope for this feature? This is really a nice keep. If it can just be integrated for pull, as simple as its in gogs will really be big. 
I am still trying to understand the use case of push mode in repos owned by a user like @yaroot said.

**Comment by NicolJiang** [comment]
need this feature, too.

**Comment by lispstudent** [comment]
Would it be enough (and healthily so) to keep repositories in sync with respective origins via a shell script on system side?

Or is there some kind of data maintained by `gitbucket` in its database that will fall out of sync?
